core/src/main/scala/cats/Traverse.scala

```diff
@@ -168,6 +168,49 @@ trait Traverse[F[_]] extends Functor[F] with Foldable[F] with UnorderedTraverse[
   def zipWithIndex[A](fa: F[A]): F[(A, Int)] =
     mapWithIndex(fa)((a, i) => (a, i))
 
+  /**
+    * Same as [[traverseWithIndexM]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def traverseWithLongIndexM[G[_], A, B](fa: F[A])(f: (A, Long) => G[B])(implicit G: Monad[G]): G[F[B]] =
+    traverse(fa)(a => StateT((s: Long) => G.map(f(a, s))(b => (s + 1, b)))).runA(0L)
+
+  /**
+    * Same as [[mapWithIndex]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def mapWithLongIndex[A, B](fa: F[A])(f: (A, Long) => B): F[B] =
+    traverseWithLongIndexM[cats.Id, A, B](fa)((a, long) => f(a, long))
+
+  /**
+    * Same as [[zipWithIndex]] but the index type is [[Long]] instead of [[Int]].
+    */
+  def zipWithLongIndex[A](fa: F[A]): F[(A, Long)] =
+    mapWithLongIndex(fa)((a, long) => (a, long))
+
+  /**
+   * If `fa` contains the element at index `idx`, 
+   * return the copy of `fa` where the element at `idx` is replaced with `b`. 
+   * If there is no element with such an index, return `None`. 
+   *
+   * The behavior is consistent with the Scala collection library's
+   * `updated` for collections such as `List`.
+   */
+  def updated[A, B >: A](fa: F[A], idx: Long, b: B): Option[F[B]] = {
+    if (idx < 0L)
+      None
+    else
+      traverse[State[Long, *], A, B](fa)(elem =>
```

Comment:
True, the `mapAccumulate` version does indeed look very similar, didn't know this function exists! I have questions about the override for `Vector` though:
1. Should I use `Vector#updated` at all? Because it seems like it takes index which is an `Int`: 
![image](https://user-images.githubusercontent.com/45356788/175093914-3cf558b4-98dd-4437-a2dc-1ea9a462e35f.png)
1. If I do use `Vector#updated`, should it be wrapped in `Try#toOption`? Should I use `Long#toInt` then?
3. If not, then how do I do it?

Reply:
To override for Vector we check to see if the index is between 0 and Int.MaxValue and if it is use Vector.updated. If not use the traverse implementation.

That said, I think Vector may fail poorly if you really try to make a Vector with tens of billions of items.